src/Network/Minio/Utils.hs

```diff
@@ -212,22 +212,13 @@ mkQuery k mv = (k,) <$> mv
 mkOptionalParams :: [(Text, Maybe Text)] -> HT.Query
 mkOptionalParams params = HT.toQuery $ uncurry  mkQuery <$> params
 
-chunkBSConduit :: (Monad m, Integral a)
-               => [a] -> C.ConduitM ByteString ByteString m ()
-chunkBSConduit s = loop 0 [] s
-  where
-    loop _ _ [] = return ()
-    loop n readChunks (size:sizes) = do
-      bsMay <- C.await
-      case bsMay of
-        Nothing -> when (n > 0) $ C.yield $ B.concat readChunks
-        Just bs -> if n + fromIntegral (B.length bs) >= size
-                   then do let (a, b) = B.splitAt (fromIntegral $ size - n) bs
-                               chunkBS = B.concat $ readChunks ++ [a]
-                           C.yield chunkBS
-                           loop (fromIntegral $ B.length b) [b] sizes
-                   else loop (n + fromIntegral (B.length bs))
-                        (readChunks ++ [bs]) (size:sizes)
+chunkBSConduit :: (Monad m) => [Int] -> C.ConduitM ByteString ByteString m ()
```

Comment:
A comment here explaining what happens when `ss` is not empty yet but the upstream in the conduit has no more data to send would be useful in future. IIUC, this is to support the case where we may be reading from a stream, whose length is not known in advance to us. We optimistically compute the part sizes for a sequential multipart upload which can accommodate up to a 5TB object.

Reply:
This function was just refactored with a better set of combinators, making it easy to understand as well. If there is no more data from upstream the conduit returns what bytes it got and stops. I will add a comment.